pkg/triggeruninstall/triggeruninstall.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package triggeruninstall
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+
+	policyv1 "open-cluster-management.io/config-policy-controller/api/v1"
+	"open-cluster-management.io/config-policy-controller/pkg/common"
+)
+
+// TriggerUninstall will add an annotation to the controller's Deployment indicating that the controller needs to
+// prepare to be uninstalled. This function will run until all ConfigurationPolicy objects have no finalizers.
+func TriggerUninstall(
+	ctx context.Context, config *rest.Config, deploymentName, deploymentNamespace, policyNamespace string,
+) error {
+	client := kubernetes.NewForConfigOrDie(config)
+	dynamicClient := dynamic.NewForConfigOrDie(config)
+
+	for {
+		klog.Info("Setting the Deployment uninstall annotation")
+		var err error
+
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context canceled before the uninstallation preparation was complete")
+		default:
+		}
+
+		deploymentRsrc := client.AppsV1().Deployments(deploymentNamespace)
+
+		deployment, err := deploymentRsrc.Get(ctx, deploymentName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		annotations := deployment.GetAnnotations()
+		annotations[common.UninstallingAnnotation] = "true"
+		deployment.SetAnnotations(annotations)
+
+		_, err = deploymentRsrc.Update(ctx, deployment, metav1.UpdateOptions{})
+		if err != nil {
+			if k8serrors.IsServerTimeout(err) || k8serrors.IsTimeout(err) || k8serrors.IsConflict(err) {
+				klog.Infof("Retrying setting the Deployment uninstall annotation due to error: %s", err)
+
+				continue
+			}
+
+			return err
```

Comment:
Why only retry with certain errors?

Reply:
I don't want the uninstall to stall out for an unrecoverable error and continuously query the Kubernetes API. The addon controller PR uses the Pod restart policy so that Kubernetes can handle restarts of such unhandled errors:
https://github.com/open-cluster-management-io/governance-policy-addon-controller/pull/60

This allows Kubernete's backoff mechanism to work and sometimes a restart could help such as when the instantiated Kubernetes client becomes stale, though unlikely. It also may alert the cluster administrator that there is a failing Pod that needs investigation.

Reply:
Got it, other retries are just handled by the container restarting.